protocol/contracts/admin/MerkleTreeClaim.sol

```diff
@@ -0,0 +1,121 @@
+//SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+/**
+ * @title A generic claim contract, with vesting, using a merkle tree to specify wallets and amounts
+ * @dev works with rebasing tokens, so can use to distribute vaulted temple
+ * @author butlerji
+ */
+contract MerkleTreeClaim is Ownable, Pausable {
+    /// @dev when the vesting starts
+    uint256 public immutable vestingStartTimestamp;
+
+    /// @dev vesting cliff duration
+    uint256 public immutable vestingCliffDuration;
+
+    /// @dev how long is the vesting period
+    uint256 public immutable vestingDuration;
+
+    /// @dev Token being claimed
+    IERC20 public immutable claimToken;
+
+    /// @dev merkle tree root of (address, claimShare)
+    bytes32 public immutable claimShareMerkleRoot;
+
+    /// @dev total shares yet to be claimed. Expect the following condition
+    /// to always hold sum(originalShares) - sum(claimedShares) == totalUnclaimedShares
+    uint256 public totalUnclaimedShares;
+
+    /// @dev shares claimed by a given account
+    mapping(address => uint256) public claimedShares;
+
+    event Claimed(address account, uint256 sharesClaimedToDate, uint256 sharesClaimedInTx, uint256 amountTokensTransferred);
+
+    constructor(
+        uint256 _vestingStartTimestamp,
+        uint256 _vestingCliffDuration,
+        uint256 _vestingDuration,
+        IERC20 _claimToken,
+        bytes32 _claimShareMerkleRoot,
+        uint256 _totalUnclaimedShares
+    ) {
+        vestingStartTimestamp = _vestingStartTimestamp;
+        vestingDuration = _vestingDuration;
+        vestingCliffDuration = _vestingCliffDuration;
+        claimToken = _claimToken;
+        claimShareMerkleRoot = _claimShareMerkleRoot;
+        totalUnclaimedShares = _totalUnclaimedShares;
+    }
+
+    /**
+     * @notice claim `share` of token `purchaseToken` for this Round. Requires valid `amount` as per `inviteCode`
+     */
+    function claimFor(address account, uint256 share, bytes32[] calldata merkleProof) external {
+        require(account != address(0), "MerkleTreeClaim: Address cannot be 0x0");
+        require(MerkleProof.verify(merkleProof, claimShareMerkleRoot, keccak256(abi.encode(account, share))), "MerkleTreeClaim: Invalid claim");
```

Comment:
I know its draft so forgive me if I'm jumping in a little early; just dumping while its on my mind. 

Is this contract intended to be reusable? I.e. to allow multiple claims? If so, there'd need to be some protection here for msg.sender (either as part of the leaf or could use eip712 to verify signature); as after the first time an address has claimed, there is nothing to prevent another address replaying using the proof from the tx data.

Reply:
Posted early to get early feedback - so no need to apologise + please jump in early :)

  1. Each contract is used for one pot of claimable tokens
  2. For any given pot, an account can claim multiple times, mainly to keep claiming during the vesting period
  3. Normally you'd have some kind of address => claim boolean, In order to have vesting, I've got an `address => claimedSoFar`  state to balance the math
  4. because it's claimFor, and the merkle tree is public kowledge anyway, the worst that can happen is I claim and send value to a given address. I can't actually take their tokens.
  
  All untested ofc - but I think your attack vector is impossible (we should think it through again before merging to master! I've added it to the checklist for this PR)